### PR TITLE
Add GCP base reference implementation

### PR DESCRIPTION
**Pull request by davidmirror-ops**
The modules added on this PR, help bootstrap a GCP environment for a successful installation of `flyte-core`. The components include:

- GKE single-region cluster, 3-nodes pool
- VPC, subnets and Private Services Connection 
- IAM policies, roles, Google Service Accounts and KSAs configured to impersonate GSAs
- GCS buckets for Terraform state and metadata
- NGINX Ingress controller
- cert-manager with local-issuer
- Reference Helm values file

 **Not included in this PR:**

- Logging
- Auth

### CODE REVIEW

**Comment by fg91** [comment]
@davidmirror-ops would you add me as reviewer? Don't have permissions in this repo.

File: environments/gcp/README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+3. Confirm the ID of your project:
+    ```bash
+    gcloud projects list
+    ```
+
+3 . Set the default project
+
+```bash
+gcloud config set project <YOUR_PROJECT_ID>
+```
```
Comment (fg91) [inline environments/gcp/README.md]: Suggestion:
My personal preference would be to not set the default project and region  via `gcloud` but instead ask the user to insert them in `locals.tf`. This way, everything is explicitly in git and there isn't suddenly a diff if another engineer runs `terraform apply` on their computer which might have a different region set.

File: environments/gcp/README.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+4. Set the default region
+    - ``gcloud config set compute/region <your-GCP-region> ``
+    - [Learn more](https://cloud.google.com/compute/docs/gcloud-compute)
+5.  Enable necessary APIs:
+    - ``gcloud services enable servicenetworking.googleapis.com container.googleapis.com compute.googleapis.com``
```
Comment (fg91) [inline environments/gcp/README.md]: Suggestion:
`container.googleapis.com` and `compute.googleapis.com` could be activated via `environments/gcp/services.tf` instead of this explicit `gcloud` command. If there is a chicken egg problem, it could be resolved with a `depends_on` in the respective resource.

File: environments/gcp/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+helm install flyte-binary flyteorg/flyte-binary --values values-gcp-binary.yaml 
+```
+5. Update the DNS A entry to point to the Ingress IP:
+    - `kubectl get ingress -n flyte`
+6. Update your `$HOME\.flyte\config,yaml` and make `endpoint` your DNS name:
+```yaml
+...
+#Example
+endpoint: dns://flyte-on-gcp.uniondemo.run 
+```
```
Comment (fg91) [inline environments/gcp/README.md]: Instead of having the user run `terraform apply`, copy the outputs into the `values-gcp-binary.yaml`, and then run `helm install`, you could do the following:

```


resource "helm_release" "flyte_binary" {
  ...
  values = [
    "${file(".../values-gcp-binary.yaml")}"
  ]
  
  # Automatically inject the values the user currently is asked to copy manually:
  set {
    name  = "userSettings.dbHost"
    value = module.flyte-db.instance_ip_address
  }

  set { ...
}
```

This way, everything might work with a single `terraform apply`.

(The only issue can arise when a `helm_release` requires the presence of CRDs in the cluster for terraform to plan. This causes a chicken egg problem but this shouldn't be the case for Flyte.)

Comment (davidmirror-ops) [reply environments/gcp/README.md]: Right! I'm working right now with `yamlencode` and `templatefile` to achieve exactly this. 

The issue, though, is that `flyte-core` is already prepared for this level of interpolation but for `flyte-binary` there are sections (like `storage`) that don't, so to achieve this for both charts, changes will have to be done there. I'll capture this on an Issue as it's ongoing work that will be ready in another PR

File: environments/gcp/gcs.tf [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  source = "terraform-google-modules/cloud-storage/google"
+
+  project_id      = local.project_id
+  location        = local.region
+  names           = ["${local.name_prefix}-data"]
+  prefix          = ""
+  set_admin_roles = true
+  admins = [
+    google_service_account.flyte_binary.member,
+    google_service_account.flyte_worker.member
```
Comment (fg91) [inline environments/gcp/gcs.tf]: Thought:
This might be a bit permissive but in the spirit of providing guidance for organizations who want to get started with flyte for the first time, I think it might make sense to keep it simple.
Users can still fine tune the permissions to their liking.

File: environments/gcp/gcs.tf [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  ]
+}
+
+resource "google_storage_bucket_iam_binding" "flyte_data_bucket_reader" {
+  bucket = module.flyte_data.name
+  role   = "roles/storage.legacyBucketReader"
```
Comment (fg91) [inline environments/gcp/gcs.tf]: Question:
Why is this role required? Does the implementation use any ACLs? `roles/storage.objectViewer` should be preferred but I wonder why the `set_admin_roles` in the module above which according to the docs does 

> Grant roles/storage.objectAdmin role to admins and bucket_admins.


<br class="Apple-interchange-newline">

is not enough 🤔 

Note:
In google cloud one needs to be careful with `iam_binding` as it means that the `members` specified here *are the only ones who have these permissions*.

If you want to add the permission to specific members but not necessarily specify all member with this permission here, you can use `iam_member` with `for_each`.

File: environments/gcp/gcs.tf [hunk 6]
```diff
@@ -0,0 +1,40 @@
+module "flyte_data" {
```
Comment (fg91) [inline environments/gcp/gcs.tf]:
```suggestion
# See Metadata in 
# https://docs.flyte.org/en/latest/concepts/data_management.html#types-of-data 
module "flyte_data" {
```

Is this correct?

File: environments/gcp/gcs.tf [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+resource "google_storage_bucket_iam_binding" "flyte_data_bucket_reader" {
+  bucket = module.flyte_data.name
+  role   = "roles/storage.legacyBucketReader"
+  members = [
+    google_service_account.flyte_binary.member,
+    google_service_account.flyte_worker.member
+  ]
+}
+
+module "flyte_user_data" {
```
Comment (fg91) [inline environments/gcp/gcs.tf]:
```suggestion
# See Raw data in 
# https://docs.flyte.org/en/latest/concepts/data_management.html#types-of-data 
module "flyte_user_data" {
```

Is this correct?

File: environments/gcp/gke.tf [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+  ip_range_services        = local.gke_services_range_name
+  create_service_account   = true
+  remove_default_node_pool = true
+
+  node_pools = [
+    {
+      name         = "default"
+      machine_type = "e2-standard-4"
+      min_count    = 0
+      max_count    = 3
```
Comment (fg91) [inline environments/gcp/gke.tf]: Suggestion

```suggestion
      max_count    = 3
      # Activate image streaming https://cloud.google.com/kubernetes-engine/docs/how-to/image-streaming to speed up pulling of images
      enable_gcfs  = false  
```

Let's not activate this by default (default is false) but let's point users to this feature, it is amazing. Our images are several GB large typically and instead of pulling them for several minutes, it takes <10 seconds. This makes for a super smooth Flyte experience.

File: environments/gcp/iam.tf [hunk 9]
```diff
@@ -0,0 +1,64 @@
+resource "google_iam_workload_identity_pool" "workload-identity-pool" {
```
Comment (fg91) [inline environments/gcp/iam.tf]: Suggestion:

Maybe we can simply configure this in the gke cluster module instead, see `identity_namespace` in the [docs](https://registry.terraform.io/modules/terraform-google-modules/kubernetes-engine/google/latest)?

(Disclaimer: have never used this module but in a raw `google_container_cluster` resource I can confirm the `workload_identity_config.workload_pool` does the trick.)

File: environments/gcp/iam.tf [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  workload_identity_pool_id = "${local.project_id}"
+}
+#add this
+
+
+resource "google_service_account" "flyte_binary" {
+  account_id = "${local.name_prefix}-flyte-binary"
+}
+
+data "google_iam_policy" "flyte_binary_workload_identity" {
```
Comment (fg91) [inline environments/gcp/iam.tf]: Thought:
In GCP, IAM policies are very authorative and one can easily take away permissions by other members by accident. Especially dangerous on the project level where one can easily [lock onself out](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam#google_project_iam_policy) 😆 

Since this is not on the project level, it might not be thaat bad but I personally would have created multiple [`google_service_account_iam_member`](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_service_account_iam#google_service_account_iam_member) resources instead.

Comment (fg91) [reply environments/gcp/iam.tf]: Feel free to just ignore/resolve!

File: environments/gcp/ingress.tf [hunk 11]
```diff
@@ -0,0 +1,77 @@
+module "nginx-controller" {
+  source  = "terraform-iaac/nginx-controller/helm"
+  version = "2.3.0"
+  namespace = "kube-system"
```
Comment (fg91) [inline environments/gcp/ingress.tf]: I would put this into its own namespace.

File: environments/gcp/ingress.tf [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+data "http" "manifestfile" {
+  url = "https://github.com/cert-manager/cert-manager/releases/download/v1.13.2/cert-manager.crds.yaml"
+}
+
+data kubectl_file_documents "cert-manager-manifest" {
+  content = data.http.manifestfile.response_body 
+
+}
+
+#Installing cert-manager CRDs from the manifest instead of using Helm is the recommended approach for production environments.
```
Comment (fg91) [inline environments/gcp/ingress.tf]: Do I remember correctly that this is to avoid a chicken-egg problem with `terraform apply`?

File: environments/gcp/ingress.tf [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+#Installing cert-manager CRDs from the manifest instead of using Helm is the recommended approach for production environments. 
+#Learn more https://cert-manager.io/docs/installation/helm/#option-1-installing-crds-with-kubectl
+
+resource kubectl_manifest "cert-manager-crds" {
+ for_each = data.kubectl_file_documents.cert-manager-manifest.manifests
+ yaml_body = each.value
+depends_on = [ module.gke ]
+}
+
+resource kubernetes_namespace "cert-manager-ns" {
```
Comment (fg91) [inline environments/gcp/ingress.tf]: Minor:
Could be created directly in the helm release resource via `create_namespace`.

File: environments/gcp/locals.tf [hunk 14]
```diff
@@ -0,0 +1,13 @@
+locals {
+  application            = "flyte"
+  environment            = "terraform"
+  name_prefix            = "${local.application}-${local.environment}"
+  region                 = data.google_client_config.current.region
+  project_id             = "<my-GCP-project>" #Insert your GCP Project ID
+  workload_identity_pool = "${local.project_id}"
```
Comment (fg91) [inline environments/gcp/locals.tf]: I think this could be shielded from the users if we just activate it in the module configuring the gke cluster.

File: environments/gcp/provider.tf [hunk 15]
```diff
@@ -0,0 +1,31 @@
+provider "google" {
+  project = "flyte-terraform1-403721" #Replace with your project name
+  region  = "us-east1"
+}
```
Comment (fg91) [inline environments/gcp/provider.tf]: I would use the values the user adapted in `locals.tf`. This way, the user one has to modify a single file.

File: environments/gcp/services.tf [hunk 16]
```diff
@@ -0,0 +1,14 @@
+locals {
+  enabled_services = [
+    "servicenetworking.googleapis.com", 
+     "container.googleapis.com",
+     "compute.googleapis.com"
```
Comment (fg91) [inline environments/gcp/services.tf]:
```suggestion
    "container.googleapis.com",
    "compute.googleapis.com"
```

File: environments/gcp/sql.tf [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    ipv4_enabled        = false,
+    enable_private_path_for_google_cloud_services = true
+    private_network     = module.network.network_self_link
+    
+    require_ssl         = null
+  }
+}
+
+
+
```
Comment (fg91) [inline environments/gcp/sql.tf]:
```suggestion
```

File: environments/gcp/terraform.tf [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    }
+    helm = {
+      source = "hashicorp/helm"
+      version = ">=2.11.0"
+  }
+
+  kubernetes = {
+      source = "hashicorp/kubernetes"
+      version = ">=2.23.0"
+    }
```
Comment (fg91) [inline environments/gcp/terraform.tf]: Is this still used? Maybe I'm overlooking the respective resource.

Personal opinion:
I prefer this provider over the `kubectl` provider and use [`k8s2tf`](https://github.com/sl1pm4t/k2tf) to convert raw manifests to resources for the kubernetes provider.
Only reason for me to use the `kubectl` provider would be to avoid the  [chicken egg problem](https://www.reddit.com/r/Terraform/comments/uwn8m9/do_you_use_the_kubernetes_terraform_provider/) terraform unfortunatly suffers from when either the cluster doesn't exist yet or doesn't have certain CRDs installed at `terraform plan` time.
Does the chicken and egg problem occur here when trying to create the cluster, install cert manager, and create the `Issuer` resource at the same time?

File: environments/gcp/terraform.tf [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+
+    http = {
+      source = "hashicorp/http"
+      version = ">=3.4.0"
+    }
+  }
+  required_version = ">= 1.3.0"
+  
+  backend "gcs" {
+    bucket = "flyte-tf-state" #Replace with the GCS Bucket you created for TF State storage
```
Comment (fg91) [inline environments/gcp/terraform.tf]: Does it work to let the user configure this in `locals.tf`?

File: environments/gcp/tfstate.tf [hunk 20]
```diff
@@ -0,0 +1,15 @@
+locals {
```
Comment (fg91) [inline environments/gcp/tfstate.tf]: Does this require multiple terraform inits/applies, one to create the bucket, the next one to configure terraform to use this bucket, and then to create the rest of the resources?

Might be very confusing for users who have never used terraform :/

Comment (davidmirror-ops) [reply environments/gcp/tfstate.tf]: I need to test this one more time before getting rid of it. It effectively creates another bucket for tfstate which is not used at all.

File: environments/gcp/terraform.tf (hunk 19 above)
Comment (davidmirror-ops) [reply environments/gcp/terraform.tf]: Done

File: environments/gcp/terraform.tf (hunk 18 above)
Comment (davidmirror-ops) [reply environments/gcp/terraform.tf:25]: I will create a new Issue to track this

File: environments/gcp/gcs.tf (hunk 4 above)
Comment (davidmirror-ops) [reply environments/gcp/gcs.tf]: Right, but still I'll investigate what would be the minimum required perms bc I think is easier for orgs to relax policies than to strengthen them, especially without having enough context.

File: environments/gcp/ingress.tf (hunk 12 above)
Comment (davidmirror-ops) [reply environments/gcp/ingress.tf]: Partially yes, the cert-manager CRDs need to be present beforehand, and creating them through a manifest is the recommended approach as opposed to using Helm, which has a number of limitations (https://cert-manager.io/docs/installation/helm/#crd-considerations)

File: environments/gcp/terraform.tf (hunk 18 above)
Comment (davidmirror-ops) [reply environments/gcp/terraform.tf]: >Does the chicken and egg problem occur here
For the most part, yes.

`cert-manager` itself requires several CRDs available before installation time, at least, for production environments. Trying to enable a `depends_on` with the GKE cluster while using `kubernetes_manifest` leads to a situation where the kubeconfig needs to be populated from an existing GKE cluster. This, in turn, can lead to inconvenient hacks like having two planning phases (one to build the cluster and another to apply the resources to it). 

In summary, that provider is effectively not being used and will be removed.

File: environments/gcp/gcs.tf (hunk 5 above)
Comment (davidmirror-ops) [reply environments/gcp/gcs.tf]: I tried to cover this on an Issue: https://github.com/unionai-oss/deploy-flyte/issues/8

File: environments/gcp/README.md (hunk 3 above)
Comment (davidmirror-ops) [reply environments/gcp/README.md]: Completed

File: environments/gcp/terraform.tf (hunk 18 above)
Comment (davidmirror-ops) [reply environments/gcp/terraform.tf]: It turns out it's been used to create a ns and a secret. A further step would be just submitting a single manifest to do both and getting rid of this provider. Leaving it for a further release (minor effort)

File: environments/gcp/gcs.tf (hunk 4 above)
Comment (davidmirror-ops) [reply environments/gcp/gcs.tf]: I found this was not necessary, nor was the `set_admin` and the bucket IAM binding as the roles granted to the GSAs allows KSA the required access to buckets using impersonation

**Comment by davidmirror-ops** [comment]
Closes #8

File: environments/gcp/flyte-core/services.tf [hunk 21]
```diff
@@ -0,0 +1,14 @@
+locals {
+  enabled_services = [
+    "servicenetworking.googleapis.com", 
+    "container.googleapis.com",
+    "compute.googleapis.com"
```
Comment (NikitaSemenovAiforia) [inline environments/gcp/flyte-core/services.tf:5]:
```suggestion
    "compute.googleapis.com",
    "sqladmin.googleapis.com"
```

I'm using Pulumi, not Terraform. So Database creation is slightly different and this service should be enabled. Or maybe I'm wrong =) anyway, that helped me.

Comment (davidmirror-ops) [reply environments/gcp/flyte-core/services.tf:5]: @NikitaSemenovAiforia Interesting, are you operating the DB with a Service Account? That would be better than local PG users (as it's done here) and would explain why you need the SQL Admin API.

I've been circulating the idea to build a ref implementation with the Pulumi Python SDK. Anything you are able to share, would be great!

Comment (NikitaSemenovAiforia) [reply environments/gcp/flyte-core/services.tf:5]: Yes, I'm using a Service Account to operate with the GCP. Pulumi invokes resources via Cloud SQL REST API under this account. That is why it should be enabled.

Great. I've just started. Will share what I got at some point

File: environments/gcp/flyte-core/sql.tf [hunk 22]
```diff
@@ -0,0 +1,52 @@
...
+    {
+      name      = "flyte"
+      charset   = ""
+      collation = ""
+    }
+  ]
+
+  additional_users = [
+    {
+      name            = "flyte"
```
Comment (NikitaSemenovAiforia) [inline environments/gcp/flyte-core/sql.tf]: I caught an authentication error. But if changed to "flyteadmin" then it works. Same with DB name. I don't know if it is necessary to create two accounts, but here "flyteadmin" is used.
https://github.com/davidmirror-ops/deploy-flyte/blob/main/environments/gcp/flyte-core/values-gcp-core.yaml#L210

Comment (NikitaSemenovAiforia) [reply environments/gcp/flyte-core/sql.tf]: Error in DB was:
`db=flyteadmin,user=flyteadmin DETAIL:  Role "flyteadmin" does not exist.`
`db=flyteadmin,user=flyteadmin FATAL:  password authentication failed for user "flyteadmin""`

Comment (davidmirror-ops) [reply environments/gcp/flyte-core/sql.tf]: Right! Sorry, I didn't port the change upstream. It's now fixed. Thanks!

File: environments/gcp/flyte-core/README.md [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+| Key      | Value |Notes |
+| ----------- | ----------- |-----|
+| `application`      | Use your own/leave default      |    This is just a label  |
+| `environment`  | Use your own/leave default    |  This is just a label    |
+| `project_id` | your GPC project ID |
+`dns-domain` | A DNS domain you own, so SSL certificates can be generated|
+|`region` | The GCP region you'll use |
+
+5. Save your changes.
+6. Go to `terraform.tf` and replace the name of the GCS bucket you created in step 2 in the appropiate section:
```
Comment (fg91) [inline environments/gcp/flyte-core/README.md]: Did you test whether this can be changed in `locals.tf` as well?

File: environments/gcp/flyte-core/README.md [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+NAME              CLASS    HOSTS                     ADDRESS         PORTS     AGE
+flyte-core        <none>   flyteontf.uniondemo.run   35.237.42.230   80, 443   3m1s
+flyte-core-grpc   <none>   flyteontf.uniondemo.run   35.237.42.230   80, 443   3m1s
+```
+3. Create a DNS `A` record in a zone you own, pointing to the Ingress IP.
+4. Update your `$HOME\.flyte\config,yaml` and make `endpoint` your DNS name:
+```yaml
+...
+#Example
+endpoint: dns://flyteontf.uniondemo.run
```
Comment (fg91) [inline environments/gcp/flyte-core/README.md]: For some reason we need three forward slashes in our config: `dns:///`

File: environments/gcp/flyte-core/README.md [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+#Example
+endpoint: dns://flyteontf.uniondemo.run 
+```
+
+> NOTE: this is only needed for CLI access (`flytectl` or `pyflyte`)
+
+### Testing your deployment
+
+
+5. In your browser, go to `https://<your-DNS-record>/console`
```
Comment (fg91) [inline environments/gcp/flyte-core/README.md]: Will auth be added in a future PR? What do you think of adding a warning here until then that this would leave your new flyte deployment exposed to the public internet and that one needs to follow the docs to activate auth until this is included? I suspect the first users will immediately deploy as soon as this is merged.

File: environments/gcp/flyte-core/flyte.tf [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+   hostname = local.flyte-host
+   flyteadminServiceAccount = google_service_account.flyteadmin_gsa.account_id
+   flytepropellerServiceAccount = google_service_account.flytepropeller_gsa.account_id
+   flyteschedulerServiceAccount = google_service_account.flytescheduler_gsa.account_id
+   datacatalogServiceAccount = google_service_account.datacatalog_gsa.account_id
+   flyteworkersServiceAccount = google_service_account.flyteworkers_gsa.account_id
+  }
+  
+  
+
```
Comment (fg91) [inline environments/gcp/flyte-core/flyte.tf]:
```suggestion
```

File: environments/gcp/flyte-core/flyte.tf [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+   flytepropellerServiceAccount = google_service_account.flytepropeller_gsa.account_id
+   flyteschedulerServiceAccount = google_service_account.flytescheduler_gsa.account_id
+   datacatalogServiceAccount = google_service_account.datacatalog_gsa.account_id
+   flyteworkersServiceAccount = google_service_account.flyteworkers_gsa.account_id
+  }
+  
+  
+  
+  )
+  ]
```
Comment (fg91) [inline environments/gcp/flyte-core/flyte.tf]: Formatting looks off, maybe run `terraform fmt` on these files.

File: environments/gcp/flyte-core/iam.tf [hunk 28]
```diff
@@ -0,0 +1,174 @@
...
+# IAM Policy for the Flyte workers role. The SA IAM Policy resource is authoritative for, in this case, the Flyteworkers SA. 
+#It is implemented in this way both as a convenience to be able to configure SA impersonation for the `default` SA of 
+#each and every project-domain combination in a programmatic way, but also because the Flyteworkers SA should only act as a
+#Workload Identity User. 
+
+data "google_iam_policy" "flyte_worker_workload_identity" {
+  binding {
+    members = formatlist("serviceAccount:${module.gke.identity_namespace}[%s]", local.flyte_worker_wi_members)
+    role    = "roles/iam.workloadIdentityUser"
+  }
```
Comment (fg91) [inline environments/gcp/flyte-core/iam.tf:167]: I haven't used this syntax to configure workload identities. I would have expected the a `service_account_id` arg here as in the `google_service_account_iam_binding` resources above.

Not saying this doesn't work, just sanity check please.

File: environments/gcp/flyte-core/ingress.tf [hunk 29]
```diff
@@ -0,0 +1,82 @@
+resource kubernetes_namespace "flyte-ns" {
+  metadata {
+  name = "flyte"
+  }
+depends_on = [ module.gke ]
+}
```
Comment (fg91) [inline environments/gcp/flyte-core/ingress.tf:6]:
```
resource helm_release "flyte-core" {
  namespace = "flyte"
  create_namespace = true
```

Is this required given that we create the namespace with the `helm_release`?

File: environments/gcp/flyte-core/ingress.tf
Comment (fg91) [inline environments/gcp/flyte-core/ingress.tf:1]: Minor:
In the other `.tf` files you named the resources with `_` instead of `-` which also is more common. Maybe change in this one here.
Optional of course.

File: environments/gcp/flyte-core/ingress.tf [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+     yaml_body = (<<-YAML
+apiVersion: cert-manager.io/v1
+kind: Issuer
+metadata:
+  name: letsencrypt-production
+  namespace: flyte
+spec:
+  acme:
+    server: https://acme-v02.api.letsencrypt.org/directory
+    email: noreply@flyte.org
```
Comment (fg91) [inline environments/gcp/flyte-core/ingress.tf]: Maybe this should be exposed via `locals.tf` as well as it should be the user's mail?

File: environments/gcp/flyte-core/ingress.tf [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+depends_on = [ module.gke ]
+}
+
+
+resource kubernetes_secret "flyte-tls-secret" {
+  metadata {
+   name = "flyte-secret-tls"
+   namespace = "flyte"
+
+  }
```
Comment (fg91) [inline environments/gcp/flyte-core/ingress.tf:45]: Why is this required? Doesn't cert manager create the secret automatically?

File: environments/gcp/flyte-core/locals.tf [hunk 32]
```diff
@@ -0,0 +1,19 @@
...
+  dns-domain =           "uniondemo.run" #Change to your domain name 
+  region   =             "us-east1" #Change to your GCP region
+
+  # Change this only if you need to add more projects in the default installation
+  flyte_projects         = ["flytesnacks"]
+  flyte_domains          = ["development", "staging", "production"]
+  
+  #Don't change this
+  flyte_ksas             = ["default"] 
+  name_prefix            = "${local.application}-${local.environment}"
```
Comment (fg91) [inline environments/gcp/flyte-core/locals.tf]: If the user isn't supposed to change this, we could put these two locals into one of the other files where they are consumed.

File: environments/gcp/flyte-core/services.tf [hunk 33]
```diff
@@ -0,0 +1,14 @@
+locals {
```
Comment (fg91) [inline environments/gcp/flyte-core/services.tf:1]: Can you please check whether cloud sql and storage need to be activated here as well in case the user starts with a fresh gcp project?

File: environments/gcp/flyte-core/values-gcp-core.yaml [hunk 34]
```diff
@@ -0,0 +1,391 @@
...
+    limits:
+      ephemeral-storage: 200Mi
+    requests:
+      cpu: 50m
+      ephemeral-storage: 200Mi
+      memory: 200Mi
+  service:
+    annotations:
+      # Required for the ingress to properly route grpc traffic to grpc port
+      cloud.google.com/app-protocols: '{"grpc":"HTTP2"}'
```
Comment (fg91) [inline environments/gcp/flyte-core/values-gcp-core.yaml]: Are you sure this is required for the nginx ingress? As far as I know, this is only needed for the gce ingress.

File: environments/gcp/flyte-core/values-gcp-core.yaml [hunk 35]
```diff
@@ -0,0 +1,391 @@
...
+      cpu: 500m
+      ephemeral-storage: 2Gi
+    requests:
+      cpu: 50m
+      ephemeral-storage: 2Gi
+      memory: 200Mi
+  service:
+    annotations:
+      # Required for the ingress to properly route grpc traffic to grpc port
+      cloud.google.com/app-protocols: '{"grpc":"HTTP2"}'
```
Comment (fg91) [inline environments/gcp/flyte-core/values-gcp-core.yaml]: Same question as for flyteadmin.

File: environments/gcp/flyte-core/README.md (hunk 24 above)
Comment (fg91) [reply environments/gcp/flyte-core/README.md]: With two, I get:

```console
    with create_connection(addr, timeout=timeout) as sock:
  File "/Users/user/.pyenv/versions/3.10.7/lib/python3.10/socket.py", line 824, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/Users/user/.pyenv/versions/3.10.7/lib/python3.10/socket.py", line 955, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 8] nodename nor servname provided, or not known
```

**Review by fg91 — CHANGES_REQUESTED** [review]
Nice that you switched to `flyte-core` over `binary` :)

Most of my comments are optional/sanity checks.

For me, the required comments to resolve are [this](https://github.com/unionai-oss/deploy-flyte/pull/7#discussion_r1401641807) and [this](https://github.com/unionai-oss/deploy-flyte/pull/7#discussion_r1401642577).

Finally, the formatting is a bit of in some of the terraform files, let's maybe run `terraform fmt` on them.

File: environments/gcp/flyte-core/README.md (hunk 25 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/README.md]: >Will auth be added in a future PR? 
Yes, #9 already captures this need. 
I'm adding a warning here as it's true it's a major risk.

File: environments/gcp/flyte-core/ingress.tf (hunk 29 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/ingress.tf:6]: We need the `flyte-secret-tls` Secret to be present in the ns before the release is installed. In theory, the cert-manager controller should still reconcile even if, after a first run, it doesn't find the secret. If we set the Helm release to depend on the secret it could create a chicken-egg problem with the namespace creation.

File: environments/gcp/flyte-core/values-gcp-core.yaml (hunk 35 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/values-gcp-core.yaml]: Right, not needed for NGINX

File: environments/gcp/flyte-core/README.md (hunk 23 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/README.md]: Yep, I tried but it errors out, it doesn't allow variables there

File: environments/gcp/flyte-core/services.tf (hunk 33 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/services.tf:1]: Tested again on two fresh projects and neither cloudsql or storage services need to be activated by the user

File: environments/gcp/flyte-core/ingress.tf (hunk 30 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/ingress.tf]: Done

File: environments/gcp/flyte-core/flyte.tf (hunk 27 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/flyte.tf]: Done!

File: environments/gcp/flyte-core/gcs.tf [hunk 36]
```diff
@@ -1,20 +1,24 @@
...
+}
+
+module "flyte-data" {
   # Metadata bucket. Learn more: https://docs.flyte.org/en/latest/concepts/data_management.html#types-of-data 
   source = "terraform-google-modules/cloud-storage/google"
 
   project_id      = local.project_id
   location        = local.region
-  names           = ["${local.name_prefix}-data"]
+  names           = ["${local.name_prefix}-data-${local.project_number}"]
```
Comment (fg91) [inline environments/gcp/flyte-core/gcs.tf:11]: Good catch.

File: environments/gcp/flyte-core/iam.tf [hunk 37]
```diff
@@ -1,3 +1,7 @@
+locals {
```
Comment (fg91) [inline environments/gcp/flyte-core/iam.tf:1]: Minor nit-pick:
I personally thick using underscores vs hyphens is more common in terraform.

File: environments/gcp/flyte-core/ingress.tf [hunk 38]
```diff
@@ -33,10 +33,11 @@ data kubectl_file_documents "cert-manager-manifest" {
 resource kubectl_manifest "cert-manager-crds" {
```
Comment (fg91) [inline environments/gcp/flyte-core/ingress.tf:33]: Nit pick:
probably not needed for the CRDs themselves but only the actual certificate later. Doesn't do any harm though.

File: environments/gcp/flyte-core/README.md [hunk 39]
```diff
@@ -88,11 +88,22 @@ flyte-core        <none>   flyteontf.uniondemo.run   35.237.42.230   80, 443   3
...
 ...
 #Example
 endpoint: dns:///flyteontf.uniondemo.run 
+insecure: false #it means, the connection uses SSL, even if it's a temporary cert-manager cert.
+
+#Uncomment only if you want to test CLI commands and the certificate is not generated yet.
+# You can confirm the cert by either going to the UI (a valid certificate should be used) or
+#from your terminal: kubectl get challenges.acme.cert-manager.io -n flyte (there should not be any pending challenge)
+
+#insecureSkipVerify: true
```
Comment (fg91) [inline environments/gcp/flyte-core/README.md:106]: Minor:
We could explain here, that with this option, SSL is used but the certificate is not checked. Am I mistaken that with `insecureSkipVerify: false` we would also have to provide a path to the respective `cert.pem`?

File: environments/gcp/flyte-core/iam.tf [hunk 40]
```diff
@@ -118,6 +118,15 @@ resource "google_project_iam_binding" "flyteworkers-binding" {
   
 }
 
+#This particular binding enables the necessary permissions to build and push images to an Artifact Registry.
```
Comment (fg91) [inline environments/gcp/flyte-core/iam.tf]: Why do the flyte workers need permission to write images into ar? Can there be a situation where images are attempted to be built in the cluster when using `ImageSpec`? Ever since GKE replaced docker in ContainerOS with containerd, it's not easily possible to build images in a GKE cluster ..

I would expect that images are always built either in CI or locally on the users notebook/dev machine, meaning that their user account needs the permission to push, not the flyte workers service account.

**Review by fg91 — APPROVED** [review]
I don't understand [this](https://github.com/unionai-oss/deploy-flyte/pull/7/commits/8209aaf69c07cc1162af51c2a965b90477646337#r1413039774) change, would be interested what the use case is.

That being said, this LGTM, awesome that this now exists, this will make Flyte adoption A LOT easier for GCP users 🚀🧑‍🚀 🧑‍🚀 🧑‍🚀

File: environments/gcp/flyte-core/iam.tf (hunk 40 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/iam.tf]: Not needed. This was being used both as the SA where users could get a token from to push images and for the Pods to pull them. This is not good. I just tested and changed the structure so there's a GSA that users/CI can leverage to obtain a token, auth to the docker repo and push images. The workers GSA is granted the `artifactregistry.reader` role to pull images.

File: environments/gcp/flyte-core/README.md (hunk 39 above)
Comment (davidmirror-ops) [reply environments/gcp/flyte-core/README.md:106]: Good question, I haven't been required to provide the cert to use it. Nevertheless, it seems to be the case (according to https://github.com/flyteorg/flytekit/blob/5c6802c97388c40d0431e4fc308e2ed651528952/flytekit/clients/auth_helper.py#L159C8-L159C8)
